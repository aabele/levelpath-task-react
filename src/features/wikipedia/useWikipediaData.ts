import React from 'react';

export enum LOADING_STATUS {
  IDLE,
  LOADING,
  SUCCESS,
  FAIL,
}

const zfill = (data: any) => String(data).padStart(2, '0');

const BASE_API = 'https://en.wikipedia.org/api/rest_v1/feed/onthisday/all';
// https://en.wikipedia.org/api/rest_v1/feed/onthisday/all/01/01

export const getDateUrl = (date: Date = new Date()): string => {
  const day = zfill(date.getDate());
  const month = zfill(date.getMonth() + 1);
  return `${BASE_API}/${month}/${day}`;
}

export interface ResponsePage {
  thumbnail?: {
    source: string;
  };
  extract: string;
  timestamp: string;
  description: string;
  desktop?: {
    page: string;
  };
}

interface ResponsePageSet {
  text: string;
  pages: ResponsePage[];
}

export type ApiResponse = {[index: string]: ResponsePageSet[]};

interface Props {
  onError(): void;
}

interface HookResponse {
  sync(): void;
  status: LOADING_STATUS;
  data: ApiResponse | null;
}

export const useWikipediaData = ({
  onError,
}: Props): HookResponse => {

  const [status, setStatus] = React.useState<LOADING_STATUS>(LOADING_STATUS.IDLE);
  const [apiResponse, setApiResponse] = React.useState<ApiResponse | null>(null);
  const [performRequest, setPerformRequest] = React.useState<boolean>(false);

  React.useEffect(() => {
    if (performRequest) {
      setStatus(LOADING_STATUS.LOADING);
      fetch(getDateUrl())
        .then((response: Response) => response.json())
        .then((response: ApiResponse) => {
          setStatus(LOADING_STATUS.SUCCESS);
          setApiResponse(response);
          setPerformRequest(false);
        }).catch(() => {
          setStatus(LOADING_STATUS.FAIL);
          setPerformRequest(false);
          onError();
        })
    }
  // eslint-disable-next-line
  }, [performRequest]);

  return {
    sync: () => setPerformRequest(true),
    status,
    data: apiResponse,
  };

}
