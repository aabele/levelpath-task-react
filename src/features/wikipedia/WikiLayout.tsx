import React, { PropsWithChildren } from 'react';
import { Button, Layout, Spin, Card } from 'antd';
import dayjs from 'dayjs';

import { ErrorModal } from './ErrorModal';
import { WikiItem } from './WikiItem';
import { LOADING_STATUS, ResponsePage, useWikipediaData } from './useWikipediaData';
import { WikiMenu } from './WikiMenu';

const { Content } = Layout;

const pageSorter = (a:ResponsePage, b: ResponsePage) => dayjs(b.timestamp).diff(dayjs(a.timestamp));

export const WikiLayout: React.FC<PropsWithChildren> = () => {
  const [isModalOpen, setIsModalOpen] = React.useState<boolean>(false);
  const onError = () => setIsModalOpen(true);
  const { data, status, sync } = useWikipediaData({ onError });
  const [activeCategory, setActiveCategory] = React.useState<string>('');
  const keys = Object.keys(data || {});

  React.useEffect(() => {
    if (keys.length) {
      setActiveCategory(keys[0]);
    }
  }, [keys.join('')]);

  const renderMenu = () => (status === LOADING_STATUS.SUCCESS && (
    <WikiMenu
      items={keys}
      setActive={setActiveCategory}
      activeKey={activeCategory}
    />
  ));

  const sortedItems: ResponsePage[] = [];

  ((data || {})[activeCategory] || [])
    .forEach((item) => {
      item.pages.forEach(page => {
        sortedItems.push(page)
      })
    });

  const visibleStatuses = [
    LOADING_STATUS.IDLE,
    LOADING_STATUS.LOADING,
  ];

  return (
    <Layout className="layout">
      <ErrorModal isOpen={isModalOpen} close={() => { setIsModalOpen(false)}} />
      {renderMenu()}
      <Content
        style={{
          padding: visibleStatuses.includes(status) ? '50px' : '0 50px',
          textAlign: visibleStatuses.includes(status) ? 'center' : 'unset',
        }}
      >
        {status === LOADING_STATUS.LOADING && (<Spin />)}
        {status === LOADING_STATUS.IDLE && (
          <Button onClick={() => sync()}>Load wikipedia data</Button>
        )}
        {status === LOADING_STATUS.SUCCESS && (
          <div className="site-layout-content" style={{ padding: '20px'}}>
            {sortedItems.sort(pageSorter).map((item, index) => (
              <Card key={`c${index}`} bordered={false}>
                <WikiItem page={item} />
              </Card>
            ))}
          </div>
        )}
      </Content>
    </Layout>
  );
}
