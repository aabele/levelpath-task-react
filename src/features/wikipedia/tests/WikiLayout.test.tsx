import React from 'react';
import { render, fireEvent, act, waitFor, screen } from '@testing-library/react';
import { getDateUrl } from '../useWikipediaData';
import { WikiLayout } from '../WikiLayout';
import apiResponse from './apiResponse.json';

describe('WikiLayout', () => {
  it('should render properly', async () => {
    await act(async () => {
      const { baseElement } = await render(<WikiLayout />);
      await expect(baseElement).toMatchSnapshot();
    });
  });

  it('should trigger fetching api data on button click', async () => {

    const mockedFetch = jest.fn(() =>
      Promise.resolve({
        json: () => Promise.resolve(apiResponse),
      }),
    ) as jest.Mock

    global.fetch = mockedFetch;

    await render(<WikiLayout />);
    await expect(mockedFetch).not.toHaveBeenCalled();
    await act(async () => {

      const item = await waitFor(() => screen.getByText(/Load wikipedia data/));
      fireEvent.click(item);
    })

    await expect(mockedFetch).toHaveBeenCalledTimes(1);
    await expect(mockedFetch).toBeCalledWith(getDateUrl());
  });

});
