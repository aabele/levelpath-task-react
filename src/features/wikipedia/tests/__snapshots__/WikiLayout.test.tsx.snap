// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WikiLayout should render properly 1`] = `
<body>
  <div />
</body>
`;
